### PR TITLE
add background for TensorArray

### PR DESCRIPTION
**Pull request by Superjom**
fix: https://github.com/PaddlePaddle/Paddle/issues/4621

### CODE REVIEW

File: doc/design/tensor_array.md [hunk 1]
```diff
@@ -1,9 +1,50 @@
 # Design for TensorArray
+## Background
+Steps are one of the core concepts of RNN. In each time step of RNN, there should be several input segments, states, and output segments; all these components act like arrays, for example, call `states[step_id]` will get the state in `step_id`th time step.
```
Comment (wangkuiyi) [inline doc/design/tensor_array.md]: I'd suggest changing the first paragraph as:

> This design doc presents the necessity of a new C++ class `TensorArray`. In addition to the very simple C++ implementation
>
> ```c++
> typedef std::vector<LoDTensor*> TensorArray;
> ```
> we also need to expose it to PaddlePaddle's Python API, because users would want to use it with our very flexible operators WhileOp.  An example for your reference:

Comment (abhinavarora) [reply doc/design/tensor_array.md]: I agree with @wangkuiyi . We should introduce the `TensorArray` and then start describing its use case in the RNN.

File: doc/design/tensor_array.md [hunk 2]
```diff
@@ -1,9 +1,50 @@
...
+
+An RNN could be implemented with the following pseudo codes
```
Comment (abhinavarora) [inline doc/design/tensor_array.md]: An RNN _can_ be implemented with the following _pseudocode_

File: doc/design/tensor_array.md [hunk 3]
```diff
@@ -1,9 +1,50 @@
...
+step = 1
+seq_len = 12
+while_loop {
+   if (step == seq_len) break;
+    states[step] = sigmoid(W * states[step-1] + U * input_segments[step]);
+    output_segments[step] = states[step] // take state as output
+   step++;
+}
+```
+According to the [RNN roadmap](https://github.com/PaddlePaddle/Paddle/issues/4561), there are several different RNNs to support.
```
Comment (abhinavarora) [inline doc/design/tensor_array.md]: several different RNNs _that Paddle will eventually_ support.

File: doc/design/tensor_array.md [hunk 4]
```diff
@@ -1,9 +1,50 @@
...
+ 
+Currently, we have an RNN implementation called `recurrent_op` which takes tensor as input; it splits the input tensors into `input_segments`.
```
Comment (abhinavarora) [inline doc/design/tensor_array.md]: Currently, the basic RNN implementation supported by Paddle is the `recurrent_op` which takes tensors as input and splits them into `input_segments`

File: doc/design/tensor_array.md [hunk 5]
```diff
@@ -1,9 +1,50 @@
...
+    states[step] = sigmoid(W * states[step-1] + U * input_segments[step]);
+    output_segments[step] = states[step] // take state as output
+   step++;
+}
+```
+According to the [RNN roadmap](https://github.com/PaddlePaddle/Paddle/issues/4561), there are several different RNNs to support.
+ 
+Currently, we have an RNN implementation called `recurrent_op` which takes tensor as input; it splits the input tensors into `input_segments`. 
+
+Considering a tensor can't store variable-length sequences directly, we proposed the tensor with the level of details (`LoDTensor` for short). Segmenting the `LoDTensor` is much more complicated than splitting a tensor, that makes it necessary to refactor the `recurrent_op` with `LoDTensor` segmenting support.
```
Comment (abhinavarora) [inline doc/design/tensor_array.md]: The first line can be changes to Since a tensor cannot store variable-length sequences directly, Paddle implements the tensor with level of details (`LoDTensor` for short).

File: doc/design/tensor_array.md [hunk 6]
```diff
@@ -1,9 +1,50 @@
...
+
+In the second stage, `dynamic_recurrent_op` should be introduced to handle inputs with variable-length sequences.
```
Comment (abhinavarora) [inline doc/design/tensor_array.md]: We can reword this as: As the next step in RNN support, `dynamic_recurrent_op` should be introduced to handle inputs with variable-length sequences.

File: doc/design/tensor_array.md [hunk 7]
```diff
@@ -1,9 +1,50 @@
...
+}
+```
+According to the [RNN roadmap](https://github.com/PaddlePaddle/Paddle/issues/4561), there are several different RNNs to support.
+ 
+Currently, we have an RNN implementation called `recurrent_op` which takes tensor as input; it splits the input tensors into `input_segments`. 
+
+Considering a tensor can't store variable-length sequences directly, we proposed the tensor with the level of details (`LoDTensor` for short). Segmenting the `LoDTensor` is much more complicated than splitting a tensor, that makes it necessary to refactor the `recurrent_op` with `LoDTensor` segmenting support.
+
+In the second stage, `dynamic_recurrent_op` should be introduced to handle inputs with variable-length sequences. 
+The implementation is the same with `recurrent_op` except that **how to split the original input `LoDTensors` and outputs to get the `input_segments` and `output_segments`** .
```
Comment (abhinavarora) [inline doc/design/tensor_array.md]: This sentence is not correct grammatically. It can be written as 
The implementation is similar to `recurrent_op`. The key difference is the way **the original input `LoDTensors` and outputs are split to get the `input_segments` and the `output_segments`**

File: doc/design/tensor_array.md [hunk 8]
```diff
@@ -1,9 +1,50 @@
...
+According to the [RNN roadmap](https://github.com/PaddlePaddle/Paddle/issues/4561), there are several different RNNs to support.
+ 
+Currently, we have an RNN implementation called `recurrent_op` which takes tensor as input; it splits the input tensors into `input_segments`. 
+
+Considering a tensor can't store variable-length sequences directly, we proposed the tensor with the level of details (`LoDTensor` for short). Segmenting the `LoDTensor` is much more complicated than splitting a tensor, that makes it necessary to refactor the `recurrent_op` with `LoDTensor` segmenting support.
+
+In the second stage, `dynamic_recurrent_op` should be introduced to handle inputs with variable-length sequences. 
+The implementation is the same with `recurrent_op` except that **how to split the original input `LoDTensors` and outputs to get the `input_segments` and `output_segments`** . 
+
+In the next stage, a dynamic RNN model based on dynamic operators would be supported. Though it can't be built on `recurrent_op` or `dynamic_recurrent_op` directly, the logic about how to split a tensor or a LoD tensor and get `input_segments` is the same.
```
Comment (abhinavarora) [inline doc/design/tensor_array.md]: The second sentence should be:
Though it can't be built over `recurrent_op` or `dynamic_recurrent_op` directly, the logic behind splitting a tensor or a LoD tensor into `input_segments` remains the same.

File: doc/design/tensor_array.md [hunk 9]
```diff
@@ -1,9 +1,50 @@
...
+
+## Why `TensorArray`
+In the three different RNNs, the logic of how to split the inputs to segments, states and outputs are similar and could be shared as a separate module.
```
Comment (abhinavarora) [inline doc/design/tensor_array.md]: the logic behind splitting the inputs to segments, states and outputs is similar and can be shared in a separate module.

File: doc/design/tensor_array.md [hunk 10]
```diff
@@ -1,9 +1,50 @@
...
+
+The array of `states`, `input_segments` and `output_segments` would be exposed to users when writing a dynamic RNN model similar to the above pseudo codes. 
+
+So there should be an array-like container which might store the segments of a tensor or LoD tensor.
```
Comment (abhinavarora) [inline doc/design/tensor_array.md]: So there should be an array-like container, which can store the segments of a tensor or LoD tensor.

File: doc/design/tensor_array.md [hunk 11]
```diff
@@ -1,9 +1,50 @@
...
+**This container could store an array of tensor and provides several methods to split a tensor or a LoD tensor** ,
```
Comment (abhinavarora) [inline doc/design/tensor_array.md]: **This container could store an array of tensors and provides several methods to split a tensor or a LoD tensor**

File: doc/design/tensor_array.md [hunk 12]
```diff
@@ -1,9 +1,50 @@
...
+that's where the notion `TensorArray` comes from.
```
Comment (abhinavarora) [inline doc/design/tensor_array.md]: This is where the notion of `TensorArray` comes from.

**Comment by Superjom** [comment]
Good suggestions, I will add some real-life use cases of `pack` and `unpack` latter.

Reply to those two questions:

- the first
For a reversed RNN, the RNN operators will have a `reversed` attribute, and that will control the RNN to traverse the sequence from tail to head.
There is another way to do this, add a `reverse()` function to `TensorArray`.

- the second
TensorArray's `unpack` splits a `LoDTensor` into batches, for example, a LoD tensor's content might have 3 sequences at some level, the sequence length is 4,3,2 respectively.

```
xxxx
xxx
xx
```

after `unpack` operation, it will be split into 4 batches:

```
0      1      2     3
x      x      x      x
x      x      x
x      x
```
The batches have 3, 3, 2, 1 instances respectively.

So the **max batch size** might be the first batch's size 3, the number of batches is 4.

The `pack` operation will concatenate the batches to the original LoD-formated Tensor.

@mkliegl
